Improve the msp_displayport funciton

ligenxxxx opened the pull request:
1. The msp displayport function can be turned on or off by setting the port function (1<<24) of the series in the CLI. msp_displayport will occupy 1 of the 3 msp serial ports at most.
2. Modified ````UART_TX_BUFFER_SIZE```` to deal with a large amount of osd elements data suddenly generated to prevent data loss.
3. HeartBeat is removed from the msp_displayport function. And drawScreen is set to execute once every 4 drawOsdElement, the user should update the previously received osdElements to the screen when drawScreen.
4. ````OSD_ARTIFICIAL_HORIZON```` is removed under the current msp_displayport function, otherwise the system will freeze.
5. I wrote a simple test program in python on the PC, and the actual effect can be observed by connecting to the computer via serial port to USB. If someone wants to test I can send the exe to a personal email.

ligenxxxx committed:
Improve the msp_displayport funciton

I wrote a simple test program in python on the PC, and the actual effect can be observed by connecting to the computer via serial port to USB. If someone wants to test I can send the exe to a personal email.

DzikuVx commented on src/main/drivers/serial_uart.h:
```diff
@@ -25,6 +25,24 @@
 // increase size further.
+#ifdef USE_MSP_DISPLAYPORT
+#define UART1_RX_BUFFER_SIZE    256
+#define UART1_TX_BUFFER_SIZE    1280
```
@ligenxxxx why 1280 and not 1024? The comment above mentions that buffer should the the power of 2

ligenxxxx replied:
I would like to modify the depth to 1024, for the current mode of transmission, the loss of a small amount of data is feasible.

DzikuVx commented on src/main/io/osd.c:
```diff
@@ -2537,7 +2537,9 @@ void osdDrawNextElement(void)
     } while(!osdDrawSingleElement(elementIndex) && index != elementIndex);
 
     // Draw artificial horizon last
+#ifndef USE_MSP_DISPLAYPORT
```
@ligenxxxx why this is required? I'd like to understand

DzikuVx replied:
And I think it should be when MSP_DISPLAYPORT is enabled and running (check at runtime), not only compiled into the code.
With current approach, if you compile with `USE_MSP_DISPLAYPORT` AHI stops working even if you do not use DISPLAYPORT

ligenxxxx replied:
This osd elements can make the system less fluid, and I don't know why.

DzikuVx replied:
Then it should be disabled when MSP_DISPLAYPORT is configured and enabled, not only compiled

DzikuVx commented on src/main/io/osd.c:
```diff
@@ -3177,7 +3179,9 @@ static void osdRefresh(timeUs_t currentTimeUs)
             fullRedraw = false;
         }
         osdDrawNextElement();
+#ifndef USE_MSP_DISPLAYPORT
```
This disabled the `displayHeartbeat` when MSP_DISPLAYPORT is compiled in. It even does not have to be enabled to disable

ligenxxxx replied:
I understand what you said now.

ligenxxxx committed:
Update serial_uart.h

resize ````UART__TX_BUFFER_SIZE``` to 1024

ligenxxxx committed:
add DrawAllElement for displayDrawScreen;modify DRAW_FREQ_DENOM = 2

known issue:
when use msp_displayport to output osd,````osdDrawSingleElement(OSD_ARTIFICIAL_HORIZON)````will case caton.

ligenxxxx committed:
Update osd.c

ligenxxxx committed:
Update osd.c

ligenxxxx reviewed (commented):
> Is this something what goes into the direction of using more OSD-Elements with Sharkbyte?
> 
> If yes, I would like to support if I can, because I'm using Sharkbyte on a 7-incher with INAV.
> (Unfortunately I'm not a software guy).
> 
> If some SW needs to be tested I will support.
> 
> ligenxxxx you are from divimath isn't it?

Yes, sharkbyte is currently developing the msp displayport function to maximize compatibility with the osd of the flight controller itself. I am not sure if the current modification is feasible, whether it will affect other functions, or whether it is worth passing, I need someone to help review it.
But currently only BTFL has a complete msp displayport function. I hope that this request can be completed before sharkbyte releases a firmware version that uses the msp displayport function.
If the sharkbyte firmware with msp displayport function is generated in the testing phase, maybe I can invite you to test it together.
Of course, the improvement of this function may also be used by DJI HD.
By the way, I am not a professional software designer, I am just interested in it, learning while practicing.

wvarty commented on src/main/fc/fc_init.c:
```diff
@@ -554,6 +554,12 @@ void init(void)
+#if defined(USE_MSP_DISPLAYPORT)
+        if (!osdDisplayPort) {
+            osdDisplayPort = displayPortMspInit();
+
```
This extra line of whitespace can be removed.

wvarty commented on src/main/io/displayport_msp.c:
```diff
@@ -85,7 +82,12 @@ static int clearScreen(displayPort_t *displayPort)
 {
     uint8_t subcmd[] = { 4 };
-    return output(displayPort, MSP_DISPLAYPORT, subcmd, sizeof(subcmd));
+    if(DrawAllElement) {
```
There should be a space between `if` and the `(` bracket here.

wvarty commented on src/main/msp/msp_serial.c:
```diff
@@ -37,8 +37,12 @@
 
+#include "config/parameter_group.h"
+#include "config/parameter_group_ids.h"
+static mspPort_t mspPorts[MAX_MSP_PORT_COUNT];
```
Might be a good idea to add a line return between the `#include` block and the static object here?

wvarty commented on src/main/msp/msp_serial.c:
```diff
@@ -50,7 +54,28 @@ void resetMspPort(mspPort_t *mspPortToReset, serialPort_t *serialPort)
+
+        portConfig = findNextSerialPortConfig(FUNCTION_MSP_DISPLAYPORT);
+    }
+    
```
No need for the double line return here. Should just be one line of whitespace.

wvarty commented on src/main/msp/msp_serial.c:
```diff
@@ -534,14 +562,24 @@ int mspSerialPush(uint8_t cmd, const uint8_t *data, int datalen)
     int ret = 0;
 
+    //Do not output in climode
+    if(cliMode){
```
Should be spaces between `if` and `(` as well as between `)` and the curly brace `{`

wvarty commented on src/main/msp/msp_serial.c:
```diff
@@ -534,14 +562,24 @@ int mspSerialPush(uint8_t cmd, const uint8_t *data, int datalen)
+        }
+        
+        // only enable for msp_displayport
+        if((msp_displayport_index & (1 << mspPort->port->identifier)) == 0) {
```
Space needed here too.

wvarty commented on src/main/target/MATEKF405/target.h:
```diff
@@ -77,6 +77,7 @@
 #define SPI2_MOSI_PIN   	    PB15
 
 #define USE_OSD
+#define USE_MSP_DISPLAYPORT
```
Is this `#define` needed on other targets too?

ligenxxxx replied:
Yes. Since I tested on the target, I only added ``#define`` to this file. I am not sure if other targets also need me to maintain.

DzikuVx replied:
Then it should be added to all targets either per target or in `common.h`

ligenxxxx replied:
````common.h```` have added it.

ligenxxxx committed:
Modified the existing problems in the code style.

ligenxxxx committed:
modify osdDraw

remove ```void osdDrawNextElement()```
add ```void osdDrawAllElement()```

DzikuVx commented on CMakeLists.txt:
```diff
@@ -29,7 +29,7 @@ else()
 endif()
 
-project(INAV VERSION 2.7.0)
+project(INAV VERSION 2.6.0)
```
The changes to this file should be reverted.

DzikuVx commented on src/main/io/osd_common.c:
```diff
@@ -116,8 +116,12 @@ void osdDrawArtificialHorizon(displayPort_t *display, displayCanvas_t *canvas, c
         osdCanvasDrawArtificialHorizon(display, canvas, p, pitchAngle, rollAngle);
     } else {
 #endif
+#if defined(USE_MSP_DISPLAYPORT)
```
This is breaking change. It should depend not on build with `USE_MSP_DISPLAYPORT` but on the `FUNCTION_MSP_DISPLAYPORT` enabled on one of the ports.
Right now, it disables 
```
osdDrawPointGetGrid(&gx, &gy, display, canvas, p);
        osdGridDrawArtificialHorizon(display, gx, gy, pitchAngle, rollAngle);
```
even if displayport is not used.
It has to be changed to dynamic, not static condition

DzikuVx commented on src/main/msp/msp_serial.c:
```diff
@@ -50,7 +55,27 @@ void resetMspPort(mspPort_t *mspPortToReset, serialPort_t *serialPort)
     uint8_t portIndex = 0;
-    serialPortConfig_t *portConfig = findSerialPortConfig(FUNCTION_MSP);
+    //FUNCTION_MSP_DISPLAYPORT detect
+    serialPortConfig_t *portConfig = findSerialPortConfig(FUNCTION_MSP_DISPLAYPORT);
```
This block should be extracted into a separate function and called only when `USE_MSP_DISPLAYPORT`
Now it will run even if `USE_MSP_DISPLAYPORT ` is not defined for a target

ligenxxxx replied:
Since msp_displayport also has the function of msp port, I don't think it should be used as a separate function.

DzikuVx replied:
It is a separate port function. Yes, it might have MSP in the name and use MSP, but is misleading from the code point of view

DzikuVx commented on src/main/io/osd.c:
```diff
@@ -3245,22 +3233,19 @@ void osdUpdate(timeUs_t currentTimeUs)
-#define DRAW_FREQ_DENOM     4
-#define STATS_FREQ_DENOM    50
+#define DRAW_FREQ_DENOM     25  // 250HZ/25 = 10HZ
+#define STATS_FREQ_DENOM    50  // 250HZ/50 = 5HZ
```
Why the change? What was wrong with the previous approach?

ligenxxxx replied:
Too fast refresh speed will cause the serial port data volume to be too large, even exceeding 115200bps. I have previously tested that when enough osd is turned on at 10HZ refresh rate to cover the entire screen, the serial port will probably be occupied by 50%～60% at 115200bps.

DzikuVx commented on src/main/io/osd.c:
```diff
@@ -3122,15 +3121,8 @@ static void osdRefresh(timeUs_t currentTimeUs)
 {
     osdFilterData(currentTimeUs);
 
-#ifdef USE_CMS
```
This should not be changed. The new behavior should be added as an option, executed ONLY is displayport is in use

DzikuVx commented on src/main/io/osd.c:
```diff
@@ -3172,11 +3164,7 @@ static void osdRefresh(timeUs_t currentTimeUs)
 #ifdef USE_CMS
     if (!displayIsGrabbed(osdDisplayPort)) {
         displayBeginTransaction(osdDisplayPort, DISPLAY_TRANSACTION_OPT_RESET_DRAWING);
-        if (fullRedraw) {
```
This seems to be breaing cuyrrent analog implementation of the OSD. Displayport adaptations should be added as an option when used, not always

ligenxxxx committed:
Solved some problems

ligenxxxx committed:
add mspDisplayportAllocatePorts()

ligenxxxx committed:
modified FUNCTION_MSP_DISPLAYPORT

FUMCTION_MSP_DISPLAYPORT = 33554432

ligenxxxx committed:
Update serial.h

ligenxxxx committed:
Update serial.h

ligenxxxx committed:
..

ligenxxxx committed:
Merge branch 'master' of https://github.com/ligenxxxx/inav

ligenxxxx committed:
Update osd.c

ligenxxxx committed:
Update osd_common.c

ligenxxxx committed:
Update osd_common.c
